[IPU] Make sure `revoked` is emitted

sed-i opened the pull request:
## Issue
Closes #79 .

## Solution
- [x] add utest(s) for IPU event
- [x] return an empty dict if in relation-broken event instead of looking into relation data, which may be out of date.

## Context
- `self.ingress.url` was returning old (non-falsey) value even after relation and/or app was removed.
- Seems like relation data is not emptied out even when looking from within a relation broken event. This may be a juju feature rather than a bug though/

sed-i commented:
@PietroPasotti I am having trouble with the utest but I'm not sure if it's because of a test bug, lib bug or harness behavior:

The relation is removed with harness, but "removed" is always blank:

```
INFO     root:ingress_per_unit.py:689 unit: ipu-requirer/0, event: RelationCreatedEvent. removed: set(); changed: set()
INFO     root:ingress_per_unit.py:689 unit: ipu-requirer/0, event: RelationJoinedEvent. removed: set(); changed: set()
INFO     root:ingress_per_unit.py:689 unit: ipu-requirer/0, event: RelationChangedEvent. removed: set(); changed: {'ipu-requirer/0'}
INFO     root:ingress_per_unit.py:689 unit: ipu-requirer/0, event: RelationChangedEvent. removed: set(); changed: set()
INFO     root:ingress_per_unit.py:689 unit: ipu-requirer/0, event: RelationBrokenEvent. removed: set(); changed: set()
```

and `revoke` doesn't fire.

sed-i commented:
`{} if isinstance(event, BrokenEvent)` per @PietroPasotti's suggestion fixed the utest.

sed-i commented on lib/charms/traefik_k8s/v1/ingress_per_unit.py:
```diff
@@ -682,7 +682,7 @@ def _handle_relation(self, event: RelationEvent):
         # we calculate the diff between the urls we were aware of
         # before and those we know now
         previous_urls = self._stored.current_urls or {}  # type: ignore
```
@rbarry82 would you use `_type_convert_stored` here?

PietroPasotti commented on lib/charms/observability_libs/v0/juju_topology.py:
```diff
@@ -0,0 +1,306 @@
+# Copyright 2022 Canonical Ltd.
```
why did you add juju_topology?

sed-i replied:
It's a dependency: the new prometheus_scrape (after fetch-lib) depends on it.

PietroPasotti reviewed (approved):
looks good!

rbarry82 reviewed (approved):
Tested manually

PietroPasotti commented:
@sed-i it seems I didn't pay enough attention while reviewing, or actually, that some changes slipped in that I didn't have a chance to review as they came in after I approved: https://github.com/canonical/traefik-k8s-operator/pull/80/commits/ce32bd3d17f9166f5ef228da7dee58cbb870ed5b

either way, there's a problem with these changes:

1) we renamed `urls` to `_get_urls_from_relation_data`, which shouldn't really be a property anymore. 
2) we replaced `urls` with a function that loads in the urls from stored data, which changes some of the semantics of what's going on, enough to break integration tests in `traefik-route`. So that's a backwards-incompatible change!

sed-i commented:
> we renamed urls to _get_urls_from_relation_data, which shouldn't really be a property anymore.

Agreed. I noticed that later too. In the IPA PR it is not a property.

> we replaced urls with a function that loads in the urls from stored data... break integration tests

Hmm. The change itself still seems correct to me logically. Can you post the link to the failing test? I'm curious.

PietroPasotti commented:
It wasn't a CI failure, it was local, so no links, but essentially what happened was:
I was writing a tester charm to represent an ingress-requiring-charm like:

```python
class MyCharm(CharmBase):
    def __init__...
        self.unit.status=Blocked()
        ingress = IPURequirer()
        if ingress.is_ready(ingress.relation):
            self.unit.status=Active()
```
This test was passing before this PR got merged.
reason: is_ready() basically checks if `bool(urls)`.
before this PR, `urls` would fetch the data in real time from the relation.
after this PR, `urls` is cached and only contains data if we are processing some events.

However, events are processed AFTER the charm is initialized, so at charm-init-time, that branch wouldn't be selected.

Took me a long time to figure that one out :)
Guess I learned to write charms always 'the proper way', even though it's just a tester charm.

sed-i commented:
Interesting. I wonder why at charm init time the old `url` would be non-false-y. And if indeed it was, then was it a mistake?
